### PR TITLE
Explicit provider installation source configuration

### PR DESCRIPTION
**Pull request by apparentlymart**
Until now, the set of places where Terraform's automatic provider installer (part of `terraform init`) has looked for available providers was generated based on a set of rules inside Terraform CLI, not directly customizable by the end-user.

The goal of this PR is to allow the user to optionally override those implied defaults in the CLI config, to allow for explicit control of source locations in special situations such as when Terraform is running in automation.

For example, the following configuration achieves the same effect as setting `-plugin-dir=/tmp/example` on the `terraform init` command line, where Terraform will _only_ search in the given directory:

```hcl
provider_installation {
  filesystem_mirror {
    path = "/tmp/example"
  }
}
```

The following configuration disables all of the local directory search locations and forces all providers to be installed solely from their origin provider registries:

```hcl
provider_installation {
  direct {}
}
```

If the user knows that in-house providers are not actually available at an origin provider registry and can be installed only from a local directory, they might say something like this:

```hcl
provider_installation {
  filesystem_mirror {
    include = ["tf.example.com/awesomecorp/*"]
  }
  direct {
    exclude = ["tf.example.com/awesomecorp/*"]
  }
}
```

This also includes some syntax for specifying network-based mirrors (on HTTP servers), but the actual implementation of that will follow in a later PR.

---

This configuration structure of having a sequence of blocks of different types where the total order is significant is not something HCL 1's decoder can represent, so unlike the rest of the `cliconfig` package this part is implemented directly using HCL 1's AST API where we can have more control over how we interpret the physica configuration structure.

That has the nice side-effect of allowing us to replicate some of the strictness we'd normally expect from the HCL 2 decoder, which I did here mainly to ease the later migration of the CLI config to HCL 2 at least for this _new_ part of the CLI configuration. (The rest remains rather lax, as before.)

---

This includes an initial draft of documentation for this new feature of the CLI configuration. This is mainly intended as a placeholder for now, because there are other documentation updates pending for the new provider namespacing and installation scheme and we'll likely want to revise these docs to better complement the broader documentation once it's written.

---

While working on this I noticed that the `TF_CLI_CONFIG_FILE` environment variable we implemented a while back to allow the user to temporarily override the CLI config search location was not disabling the CLI configuration _directory_, and so it was ending up making a strange blend of overridden and standard options.

There's a commit in this set that fixes that bug, so that `TF_CLI_CONFIG_FILE` completely disables all of the normal CLI configuration file search locations. That makes `TF_CLI_CONFIG_FILE` more useful as a way to isolate wrapper script behavior from ambient CLI configuration, but it is a breaking change we'll need to note in the changelog once we merge this.

### CODE REVIEW

File: command/cliconfig/cliconfig.go [hunk 1]
```diff
@@ -91,12 +97,23 @@ func LoadConfig() (*Config, tfdiags.Diagnostics) {
...
+	if cliConfigFileOverride() == "" {
+		if configDir, err := ConfigDir(); err == nil {
+			if info, err := os.Stat(configDir); err == nil && info.IsDir() {
+				dirConfig, dirDiags := loadConfigDir(configDir)
+				diags = diags.Append(dirDiags)
+				config = config.Merge(dirConfig)
+			}
 		}
+	} else {
+		log.Printf("Not reading CLI config directory because config location is overridden by environment variable")
```
Comment (mildwonkey) [inline command/cliconfig/cliconfig.go:116]: Does this need a log level?  `"[INFO]"` (or whatever you deem appropriate) at the start of the string?

File: command/cliconfig/cliconfig.go [hunk 2]
```diff
@@ -136,6 +153,13 @@ func loadConfigFile(path string) (*Config, tfdiags.Diagnostics) {
 		return result, diags
 	}
 
+	// Deal with the provider_installation block, which is not handled using
+	// DecodeObject because its structure is not compatible with the
```
Comment (mildwonkey) [inline command/cliconfig/cliconfig.go:157]: I started to write a comment that I'd like to see more detail here for future readers (ie, me) but found it in the function instead so 🎉 thank you!

File: command/cliconfig/provider_installation.go [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+//
+// This uses the HCL AST directly, rather than HCL's decoder, because the
+// intended configuration structure can't be represented using the HCL
+// decoder's struct tags. This structure is intended as something that would
+// be relatively easier to deal with in HCL 2 once we eventually migrate
+// CLI config over to that, and so this function is stricter than HCL 1's
+// decoder would be in terms of exactly what configuration shape it is
+// expecting.
+//
+// Note that this function wants the top-level file object which might or
```
Comment (mildwonkey) [inline command/cliconfig/provider_installation.go:28]: 👏 SUCH GOOD DOCUMENTATION thank you

File: main.go [hunk 4]
```diff
@@ -168,7 +169,22 @@ func wrappedMain() int {
 	// direct from a registry. In future there should be a mechanism to
 	// configure providers sources from the CLI config, which will then
```
Comment (mildwonkey) [inline main.go:170]: I think this comment can be updated now.

File: website/docs/commands/cli-config.html.markdown [hunk 5]
```diff
@@ -144,14 +148,199 @@ To learn how to write and install your own credentials helpers to integrate
...
+Terraform allows omitting the hostname portion for providers on
+`registry.terraform.io`, so we'd normally write `hashicorp/aws` instead in
+this case.
+
+Downloading a plugin directly from its origin registry is not always
+appropriate, though. For example, the system where you are running Terraform
+may not be able to access an origin registry due to firewall restrictions
+within your organization or your locality.
+
+To allow using Terraform in these situations, there are some alternative
```
Comment (mildwonkey) [inline website/docs/commands/cli-config.html.markdown:165]: tiny nitpick (and totally ok if you don't agree): it's more specific and accurate to say this features allows installing/using terraform _providers_ in these situations

**Comment by mildwonkey** [comment]
This looks great! I left a couple of small comments, nothing that needs to block you from merging this. 

I have two larger questions about this new feature that are out of scope for this PR, but I want to write them down somewhere before I forget: 

1. Do you think we should check for (relatively) obvious conflicts in the `provider_installation` block? Something like the example below, where the include and exclude attributes are the same:

```
  network_mirror {
    url     = "https://tf-Mirror.example.com/"
    include = ["registry.terraform.io/*/*"]
    exclude = ["registry.Terraform.io/*/*"]
  }
```
I can imagine that there are too many possible permutations to make checking this for the user onerous - at the very least we can wait and see what kind of edge cases and issues pop up.

2. We've talked about the provider developer experience and a few different ideas for simplifying the process. What do you think about adding a specific attribute (exclusive to `filesystem_mirror`s) to let provider developers point directly to a binary, forgoing the need to create a directory structure?  

I'm making this up as I go and have zero attachment to the wording or syntax, but something that conveys my intent:

```
  filesystem_mirror {
    include = ["tf.example.com/awesomecorp/*"]

    # this is my awkward attempt to only use attributes instead of blocks
    exact = ["{provider_source}:{/path/to/binary}"]

    # for example:  
    exact = ["example.com/myorg/happcloud":"$HOME/go/bin"]
  }
```

We don't have to discuss these questions here!

File: command/cliconfig/cliconfig.go (hunk 1 above)
Comment (apparentlymart) [reply command/cliconfig/cliconfig.go:116]: It seems like some other log lines in this area don't tend to have it, but I suppose I don't really know why that is, so I will add levels to all the new ones I've added around here.

File: website/docs/commands/cli-config.html.markdown (hunk 5 above)
Comment (apparentlymart) [reply website/docs/commands/cli-config.html.markdown:165]: Fair! I think I was assuming that Terraform is useless if you can't install providers, but no need to be vague about it.

**Comment by apparentlymart** [comment]
@mildwonkey and I discussed the follow-on questions above in an offline discussion today. I just wanted to capture the high-level conclusions here for posterity:

1. We will not have any explicit checks for overlapping/conflicting declarations at first because the wildcard support makes that tricky to do robustly and we want to see what sorts of problems tend to arise in practice before spending the time to research and implement such checks.

2. We will be having a follow-up discussion with provider developers about a few different options for how to approximate the old provider development workflow while giving Terraform the metadata it needs to understand how to use a locally-built provider. A CLI configuration option for this is one of the ideas on the table, but it's not clear that this would be the most ergonomic answer because it could make it easy to accidentally leave temporary development configuration in your main CLI config and then be confused later when `terraform init` installs something other than what it would do by default. We'll review a few different options and see which one seems best to implement for the 0.13 release.

**Pull request merged**